scipy/optimize/_lsq/givens_elimination.pyx

```diff
@@ -44,7 +47,7 @@ def givens_elimination(double [:, ::1] S, double [:] v, double [:] diag):
         if diag[i] == 0:
             continue
 
-        diag_row[:] = 0
+        diag_row[i+1:] = 0
```

Comment:
In the loop `for j in range(i, n)` only `diag_row[j:]` is used and modified. The is idea is to sequentially zero out j-th element of `diag_row` (starting with j=i), but as this element is never used anymore I wrote `diag_row[j] is implicitly 0 now`, i.e. we don't have to set it to 0. 

So this optimization looks correct for me. And tests are passing too.

> The new and old code are only equivalent is diag_row[i] really is reset to 0

Can you explain why it seems to you like this? Maybe I'm missing something.